File: src/0042-trapping-rain-water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height or len(height) == 0: return 0

        left, right, result = 0, len(height) - 1, 0
        leftMax, rightMax = height[left], height[right]

        while left < right:
            if leftMax < rightMax:
                left += 1
                leftMax = max(leftMax, height[left])
                result += leftMax - height[left]
            else:
                right -= 1
                rightMax = max(rightMax, height[right])
                result += rightMax - height[right]

        return result

File: src/0042-trapping-rain-water/solution.java
class Solution {
    public int trap(int[] h) {
        int N = h.length, ans = 0;

        if (N == 0) return 0;

        int[] l = new int[N];
        int[] r = new int[N];

        l[0] = h[0];
        for(int i = 1; i < N; i++)
            l[i] = Math.max(l[i-1], h[i]);

        r[N - 1] = h[N - 1];
        for(int i = N - 2; i >= 0; i--)
            r[i] = Math.max(r[i+1], h[i]);

        for(int i = 0; i < N; i++)
            ans += Math.min(l[i], r[i]) - h[i];

        return ans;
    }
}
